src/units/abilities.cpp

```diff
@@ -1297,48 +1297,41 @@ static bool overwrite_special_affects(const config& special)
 {
-	bool overwrite_self = false;
-	bool overwrite_opponent = false;
-	bool overwrite_either = false;
-
-	for(const auto& i : abil_list) {
-		if((*i.ability_cfg)["overwrite_specials"] == "both_sides") {
-			overwrite_either = true;
-			break;
-		}
-		if((*i.ability_cfg)["overwrite_specials"] == "one_side" && special_active_impl(shared_from_this(), other_attack_, *i.ability_cfg, AFFECT_SELF, ability, filter_self) && !overwrite_self) {
-			overwrite_self = true;
-		}
-		if((*i.ability_cfg)["overwrite_specials"] == "one_side" && special_active_impl(other_attack_, shared_from_this(), *i.ability_cfg, AFFECT_OTHER, ability, filter_self) && !overwrite_opponent) {
-			overwrite_opponent = true;
+	for(unit_ability_list::iterator i = overwriters.begin(); i != overwriters.end();) {
+		if(!overwrite_special_affects(*i->ability_cfg)) {
+			i = overwriters.erase(i);
+		} else {
+			++i;
 		}
 	}
-	if(!overwrite_either && !overwrite_self && !overwrite_opponent){
-		return temp_list;
-	}
-
-	if(!overwrite_either && overwrite_self && overwrite_opponent){
-		overwrite_either = true;
+	if(overwriters.empty()){
+		return input;
 	}
 
-	// At this point we need to return a changed list, so create a non-const one to return
-	unit_ability_list overwrite_list;
-	for(const auto& i : temp_list) {
-		bool overwrite = false;
-		if(overwrite_either){
-			overwrite = !is_special && overwrite_special_affects(*i.ability_cfg);
-		} else if(overwrite_self){
-			overwrite = (!is_special && overwrite_special_affects(*i.ability_cfg)) || special_active_impl(other_attack_, shared_from_this(), *i.ability_cfg, AFFECT_OTHER, ability, filter_self);
-		} else if(overwrite_opponent){
-			overwrite = (!is_special && overwrite_special_affects(*i.ability_cfg)) || special_active_impl(shared_from_this(), other_attack_, *i.ability_cfg, AFFECT_SELF, ability, filter_self);
-		}
-		if(overwrite) {
-			overwrite_list.emplace_back(i);
+	for(const auto& i : overwriters) {
+		bool affect_side = ((*i.ability_cfg)["overwrite_specials"] == "one_side");
+		for(unit_ability_list::iterator j = input.begin(); j != input.end();) {
+			bool is_overwritable = (is_special || !overwrite_special_affects(*j->ability_cfg));
+			bool one_side_overwritable = true;
+			if(affect_side && is_overwritable){
+				if(special_active_impl(shared_from_this(), other_attack_, *i.ability_cfg, AFFECT_SELF, ability, filter_self)){
+					one_side_overwritable = special_active_impl(shared_from_this(), other_attack_, *j->ability_cfg, AFFECT_SELF, ability, filter_self);
+				}
+				else if(special_active_impl(other_attack_, shared_from_this(), *i.ability_cfg, AFFECT_OTHER, ability, filter_self)){
+					one_side_overwritable = special_active_impl(other_attack_, shared_from_this(), *j->ability_cfg, AFFECT_OTHER, ability, filter_self);
+				}
+			}
+			is_overwritable = is_overwritable && one_side_overwritable;
+			if(is_overwritable) {
```

Comment:
The line above could be combined with this one, so it's just `if(is_overwritable && one_side_overwritable)`. However, no need to do that if your next PR is going to add additional logic that's easier with them kept separate.